File: java/src/main/java/be/playing/with/projections/model/Event.java
package be.playing.with.projections.model;

import java.time.LocalDateTime;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

/**
 * @author Jan Van Rensbergen
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public class Event {

  private String id;
  private EventType type;
  private LocalDateTime timestamp;
  private Map<String, String> payload;

  public Event(String id, EventType type, LocalDateTime timestamp, Map<String, String> payload) {
    this.id = id;
    this.type = type;
    this.timestamp = timestamp;
    this.payload = payload;
  }

  private Event() {
    // for frameworks
  }

  public String getId() {
    return id;
  }

  public EventType getType() {
    return type;
  }

  public LocalDateTime getTimestamp() {
    return timestamp;
  }

  public Map<String, String> getPayload() {
    return payload;
  }

  @Override
  public String toString() {
    return "Event{" +
           "id='" + id + '\'' +
           ", type='" + type + '\'' +
           ", timestamp=" + timestamp +
           ", payload=" + payload +
           '}';
  }
}

File: kotlin/src/main/kotlin/be/playing/with/projections/model/Event.kt
package be.playing.with.projections.model
import com.fasterxml.jackson.annotation.JsonValue
import java.time.LocalDateTime

data class Event(val id: String, val type: EventType, val timestamp: LocalDateTime, val payload: Map<String, String>)

enum class EventType(@get:JsonValue val representation: String) {

    PLAYER_JOINED_GAME("PlayerJoinedGame"),
    PLAYER_HAS_REGISTERED("PlayerHasRegistered"),
    GAME_WAS_STARTED("GameWasStarted"),
    QUESTION_WAS_ASKED("QuestionWasAsked"),
    ANSWER_WAS_GIVEN("AnswerWasGiven"),
    QUESTION_WAS_COMPLETED("QuestionWasCompleted"),
    GAME_WAS_OPENED("GameWasOpened"),
    GAME_WAS_FINISHED("GameWasFinished"),
    QUIZ_WAS_CREATED("QuizWasCreated"),
    QUESTION_ADDED_TO_QUIZ("QuestionAddedToQuiz"),
    QUIZ_WAS_PUBLISHED("QuizWasPublished");


    companion object {
        fun ofRepresentation(representation: String): EventType = EventType.values().first { it.representation == representation }
    }
}
